### ISSUE TITLE
-Werror prevents building of bytecode compiler on Solaris SPARC-64 with GCC

### ISSUE THREAD

**Issue by johnwhitington** [ocaml/ocaml#11324]
We get:

```
gcc -c -O2 -fno-strict-aliasing -fwrapv -D_POSIX_PTHREAD_SEMANTICS -pthread -g -Wall -Werror -fno-common -fexcess-precision=standard  -I./runtime  -D_FILE_OFFSET_BITS=64  -DCAMLDLLIMPORT= -DDEBUG  -o runtime/instrtrace.bd.o runtime/instrtrace.c
In file included from runtime/instrtrace.c:32:
runtime/caml/opnames.h:6:9: error: initialization of ‘char *’ from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
    6 |   ACC0, ACC1, ACC2, ACC3, ACC4, ACC5, ACC6, ACC7,
      |         ^~~~
```

I had the customer make this manual change to configure on line 12772:

```
replace

  warn_error_flag='-Werror'

with

  warn_error_flag=''
```

That fixes it. But it would be nice to have a flag to disable warnings as errors. All I can see is:

```
  --enable-warn-error     treat C compiler warnings as errors
```

But at line 12758, it seems to get set unconditionally for GCC:

```
case $ocaml_cv_cc_vendor in #(
  xlc-*) :
    outputobj='-o $(EMPTY)'
    warn_error_flag=''
    cc_warnings='-qflag=i:i' ;; #(
  # all warnings enabled
  sunc-*) :
    outputobj='-o $(EMPTY)'; cc_warnings="" ;; #(
  msvc-*) :
    outputobj='-Fo'
    warn_error_flag='-WX'
    cc_warnings='' ;; #(
  *) :
    outputobj='-o $(EMPTY)'
  warn_error_flag='-Werror'
  cc_warnings='-Wall -Wdeclaration-after-statement' ;;
esac
```
If it's going to be set by default, surely we want a way to disable it with a flag?

**Comment by dra27** [comment]
It's not unconditional - that's merely setting a variable which says what the flag for enabling errors as warnings is. It's not actually set until a few lines later, and that's based on the version.

I hope you're compiling a development or git branch of the compiler - release tags and tarballs should disable it by default. Although autoconf doesn't display it, all the options are available in pairs so `./configure --disable-warn-error` should disable warnings-as-errors _regardless of the version_

**Comment by johnwhitington** [comment]
Thanks.

You are right - false alarm - the customer was accidentally compiling trunk, when I had asked them to download the source for the tagged 4.14.0.

**Comment by shindere** [comment]
David Allsopp (2022/06/16 08:04 -0700):
> It's not unconditional - that's merely setting a variable which says what the flag for enabling errors as warnings is. It's not actually set until a few lines later, and that's based on the version.
>
> I hope you're compiling a development or git branch of the compiler -
> release tags and tarballs should disable it by default. Although
> autoconf doesn't display it, all the options are available in pairs so
> `./configure --disable-warn-error` should disable warnings-as-errors
> _regardless of the version_

It'd still be nice to understand the warning...

**Comment by johnwhitington** [comment]
Here's the full log, by the way:

https://gist.github.com/johnwhitington/bfcc66908c8b5ebfb5f591b9b6c9fd53

**Comment by damiendoligez** [comment]
That's only the C compiler warnings/errors. What failed is the `sed` command at line 787 of `Makefile`. Rather, it failed to insert double quotes in the right places. Do you have the full `make` log ?

**Comment by johnwhitington** [comment]
@damiendoligez I will ask the customer

**Comment by damiendoligez** [comment]
Note that it isn't critical as it happens on a file that we use internally for debugging. Since we don't use Solaris to develop the compiler, we could also just ignore the problem.

**Comment by johnwhitington** [comment]
The customer is unwilling to re-run the compilation to get the full log, I'm afraid.

**Comment by shindere** [comment]
John Whitington (2022/06/16 12:51 -0700):
> The customer is unwilling to re-run the compilation to get the full
> log, I'm afraid.

Too bad. Would be nice to get an access to a similar machine to
experiment.

**Comment by johnwhitington** [comment]
I once had a free account with these people, who have many systems:

http://www.polarhome.com

The free accounts were too limited on either memory or space to compile OCaml back in the day, but that was because Camlp4 was in-tree. I suspect it might work now.

**Comment by shindere** [comment]
John Whitington (2022/06/16 13:23 -0700):
> I once had a free account with these people, who have many systems:
>
> http://www.polarhome.com
>
> The free accounts were limited on either memory or space to compile
> OCaml back in the day, but that was because Camlp4 was in-tree. I
> suspect it might work now.

Interesting! Thanks!

If people aspire to had a good support of OCaml for their OS, is it
unrealistic to think that they could provide at least a free account but
perhaps more?

**Comment by shindere** [comment]
John Whitington (2022/06/16 08:09 -0700):
> Closed #11324 as completed.

Can we try to understand this warning, please?

Just to make sure it's not a code smell...

**Comment by dra27** [comment]
For some reason, the `sed` command on L791 of `Makefile` didn't work (`s/\([[:upper:]][[:upper:]_0-9]*\)/"\1"/g`). Although that command could be slightly simplified (there's no need for the capture group), it is POSIX. It's possible that the original user had a locale setting that made something strange happen, although it's odd that apparently _none_ of `A-Z` would ever be in the upper class. However, there's not a lot we can do without a system to confirm the error on.
